### ISSUE TITLE
Support GitOps flow with fluxv2 plugin

### ISSUE THREAD

**Issue by admincasper** [vmware-tanzu/kubeapps#6747]
**Summary**
Referencing #979 since it is closed and is not a proposal. I would like kubeapps to have a fluxv2 plugin that could generate helmreleases in a git repository.
This is to support already existing gitops workflows that teams have with argocd and flux. 

**Background and rationale**
At the current state, the kubeapps flux packaging is impractical since it is in conflict with the flux workflow. 
Any changes done by kubeapps will just get overwritten by flux.

**Description**
I would like to be able to provide credentials to a git repository, for example a k8s secret. Which could contain the URL to the repository, and the path in the git repository to install the helmrelease. I.e `add-ons/namespace1/`. 

**Acceptance criteria**
- Only people who can view the helm chart in the namespace should be able to provide git url and repo path. 
- Only people who can edit secrets in the namespace should be able to provide credentials
- Options for creating a pull request, or commiting directly into the git repo.

**Additional context**
Pull request acceptance criteria could be solved with just providing the branch name. If 'main' or 'master': direct commit.
If other branch name: Direct commit there or create branch if it does not exist.

**Comment by absoludity** [comment]
> I would like kubeapps to have a fluxv2 plugin that could generate helmreleases in a git repository.

And HelmRepositories in a git repository?

More generally, this would be a compromise of gitops with a Kubeapps UX, in that the Kubeapps UX enables easier updating of your git repo for those users with RBAC, but means you're not having the security of gitops even if you have some auditability / tracking of changes (though you can't trust the user necessarily).

In the past I think I've argued that this shouldn't be something we support, given that if people are using gitops, they *should* be using git (and the built-in ssh/gpg security), not a UX that tries to make that simpler and remove the git from gitops. But this is not the first time we've had this request, so it seems the combination of ease-of-use with some auditability is wanted more broadly.

This would be a large piece of work (a separate plugin). Happy to help people who are interested to implement this too, but it will require a fair amount of go + grpc knowledge.

**Comment by admincasper** [comment]
> And HelmRepositories in a git repository?
> 

Yes, point to e.g `flux-system` namespace where HelmRepositories reside.


> More generally, this would be a compromise of gitops with a Kubeapps UX, in that the Kubeapps UX enables easier updating of your git repo for those users with RBAC, but means you're not having the security of gitops even if you have some auditability / tracking of changes (though you can't trust the user necessarily).
> 
> In the past I think I've argued that this shouldn't be something we support, given that if people are using gitops, they _should_ be using git (and the built-in ssh/gpg security), not a UX that tries to make that simpler and remove the git from gitops. But this is not the first time we've had this request, so it seems the combination of ease-of-use with some auditability is wanted more broadly.

Yes.. with gitops you usually use a pipeline to push to a manifest repository automatically. You're not skipping Git principles by using the pipeline. It depends on your implementation of permissions to the agent as you're using Git permissions. 
If the Kubeapps git agent can create branches, commit to the branches, and open pull requests, there would be minimal problems with git security. (No direct push to main), but that all depends on the permissions in the provided secret.  

This of course would mean Kubeapps has to support authentication to the big Git providers. GitLab, GitHub and our case Azure DevOps.

> This would be a large piece of work (a separate plugin). But it will require a fair amount of go + grpc knowledge.

I agree, but I can clearly see the potential of Kubeapps to simplify the administration of HelmReleases significantly. 
If one limits access to Kubeapps to only cluster administrators this would still solve a lot of problems of cluster administration.

**Comment by KlavsKlavsen** [comment]
Yes - enabling KubeApps to make changes by doing:
Git commit (on behalf of user - with notes on who requested it in git commit)
and then effect the change so f.ex. argocd - says application is in sync

Would be awesome.. a userfriendly UI - with gitops.
It could allow "opt out" of gitops -- for "testing purposes" - and just "note" that "this application is not in sync with gitops" - and offer to commit it later - if it turns out to be permanent..
